File: BackEnd/src/main/java/com/application/service/SubjectService.java
package com.application.service;

import com.application.dto.FullSubjectDTO;
import com.application.dto.SubjectDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;

public interface SubjectService {

    SubjectDTO create(SubjectDTO subject);

    SubjectDTO read(Long id);

    Page<FullSubjectDTO> readAll(PageRequest pageRequest);

    SubjectDTO update(SubjectDTO subject);

    void delete(Long id);
}

File: FrontEnd/src/app/service/subject.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {environment} from "src/environments/environment";
import {Page} from "../interface/page";
import {Subject} from "../interface/subject";
import {Util} from "../util";

@Injectable()
export class SubjectService {
  private apiBaseUrl = environment.apiBaseUrl + '/subjects';

  constructor(private http: HttpClient) {
  }

  public create(subject: Subject): Observable<Subject> {
    return this.http.post<Subject>(`${this.apiBaseUrl}`, subject);
  }

  public update(subject: Subject): Observable<Subject> {
    return this.http.put<Subject>(`${this.apiBaseUrl}`, subject);
  }

  public delete(id: number): Observable<string> {
    return this.http.delete<string>(`${this.apiBaseUrl}/${id}`);
  }

  public readAll(page: number, pageSize: number, sortBy: string, order: string): Observable<Page<Subject>> {
    if (Util.isNullOrEmpty(sortBy, order)) {
      return this.http.get<Page<Subject>>(`${this.apiBaseUrl}?page=${page}&page_size=${pageSize}`);
    }
    return this.http.get<Page<Subject>>(`${this.apiBaseUrl}?page=${page}&page_size=${pageSize}&sort_by=${sortBy}&order=${order}`);
  }
}
